### PR TITLE
Thread safety fixes

### PR DESCRIPTION
**Pull request by edenhill**

### CODE REVIEW

File: tests/test_threads.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python
+
+from confluent_kafka import Producer, KafkaError, KafkaException
+import threading, time, queue
```
Comment (ewencp) [inline tests/test_threads.py]: Hmm, `queue` is the py3 name, whereas `Queue` is used in py2. Currently this fails on py2 for me.

File: tests/test_threads.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+
+class IntendedException (Exception):
+    pass
+
+def thread_run (myid,p,q):
```
Comment (ewencp) [inline tests/test_threads.py]: these parameter lists might fail pep8 checks, normally there are spaces after commas.

File: tests/test_threads.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            cb = do_crash
+        p.produce('mytopic', value='hi', callback=cb)
+        t = time.time()
+        try:
+            p.flush()
+            print(myid, 'Flush took %.3f' % (time.time() - t))
+        except IntendedException:
+            print(myid, "Intentional callback crash: ok")
+            continue
+        except:
```
Comment (ewencp) [inline tests/test_threads.py]: This final `except` block isn't needed

File: tests/test_threads.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+def test_thread_safety():
+    """ Basic thread safety tests. """
+
+    q = queue.Queue()
+    p = Producer({'socket.timeout.ms':10,
+                  'socket.blocking.max.ms': 10,
+                  'default.topic.config': {'message.timeout.ms': 10}})
+
+    threads = list()
+    for i in range(1, 5):
```
Comment (ewencp) [inline tests/test_threads.py]: Are the settings you have here (5 threads, 3 messages each) enough to reliably trigger an error before the patch?

**Comment by ewencp** [comment]
@edenhill Patch looks good, just had questions re: test

File: tests/test_threads.py (hunk 4 above)
Comment (edenhill) [inline tests/test_threads.py]: Yup, fails immediately and repeatedly on master:
```
$ python tests/t.py 
Exception in thread 1:
NoneType

Fatal Python error: PyEval_RestoreThread: NULL tstate
```

**Pull request merged**